Python/compile.c

```diff
@@ -6883,6 +6893,60 @@ optimize_basic_block(basicblock *bb, PyObject *consts)
                             bb->b_instr[i+1].i_opcode = NOP;
                         }
                         break;
+                    case BINARY_ADD:
+                        cnt = PyList_GET_ITEM(consts, oparg);
+                        if (PyLong_CheckExact(cnt)
+                            && inst->i_lineno == bb->b_instr[i+1].i_lineno) {
+                            int ovf = 0;
+                            long val = PyLong_AsLongAndOverflow(cnt, &ovf);
+                            // TODO: What about larger values?
+                            // They would cause an EXTENDED_ARG to be generated,
+                            // which may defeat any potential cost savings.
+                            if (ovf == 0 && val >= 0 && val < 256) {
+                                inst->i_opcode = ADD_INT;
+                                inst->i_oparg = val;
+                                bb->b_instr[i+1].i_opcode = NOP;
+                                break;
+                            }
+                        }
+                        break;
+                    # if 1
+                    case RETURN_VALUE:
+                        // LOAD_CONST(None) + RETURN_VALUE -> RETURN_NONE
+                        // LOAD_CONST(other) + RETURN_VALUE -> RETURN_CONST
+                        if (inst->i_lineno == bb->b_instr[i+1].i_lineno) {
+                            cnt = PyList_GET_ITEM(consts, oparg);
+                            inst->i_opcode = RETURN_CONST;
+                                // cnt == Py_None ? RETURN_NONE : RETURN_CONST;
+                            bb->b_instr[i+1].i_opcode = NOP;
```

Comment:
This makes the last instruction of the BB a non-terminator. Either put the NOP first or shorten the BB by one.
You'll need to add RETURN_CONST as a BB terminator to all the code that cares about such things.